### PR TITLE
Bug 1679019 - Selecting the checkbox for an alert summary in perfherder should check only the untriaged alerts

### PR DESCRIPTION
**Pull request by esanuandra**

### CODE REVIEW

**Comment by esanuandra** [comment]
image

**Comment by esanuandra** [comment]
Updated version 
image

**Review by ionutgoldan — CHANGES_REQUESTED** [review]
~~We should disable the dropdown if the user isn't logged in.~~
Oh, I just noticed this is disabled. Nevermind....

dropdown-shouldnt-be-visible-if-logout

**Comment by ionutgoldan** [comment]
We should provide test coverage for this aspect as well.

File: ui/perfherder/alerts/AlertTable.jsx [hunk 1]
```diff
@@ -176,6 +169,46 @@ export default class AlertTable extends React.Component {
...
+      untriaged: 'untriaged',
+    };
+    let selectedAlerts = [...alertSummary.alerts];
+
+    if (selectedStatus === status.none) {
+      selectedAlerts = [];
+      allSelected = false;
+    }
+
+    if (selectedStatus === status.all) {
```
Comment (ionutgoldan) [inline ui/perfherder/alerts/AlertTable.jsx:189]: Could we do an if/else if chain? It's easier to assert the code's implications this way.
(For example, if the status is none, we'd hit the 1st branch & the 3rd one.)

**Review by ionutgoldan — CHANGES_REQUESTED** [review]

File: ui/perfherder/alerts/AlertTable.jsx [hunk 2]
```diff
@@ -217,32 +250,29 @@ export default class AlertTable extends React.Component {
...
-                          repoModel={repoModel}
-                          issueTrackers={issueTrackers}
-                          user={user}
-                          updateAssignee={this.updateAssignee}
-                        />
-                      </Label>
+                    <FormGroup check className="d-inline-flex">
+                      <SelectAlertsDropdown
+                        selectAlertsByStatus={this.selectAlertsByStatus}
+                        toggleAllAlerts={() => {
```
Comment (ionutgoldan) [inline ui/perfherder/alerts/AlertTable.jsx:256]: Couldn't we remove this small helper function? We seem to already do the same thing via `selectAlertByStatus(all)`.

**Review by ionutgoldan — CHANGES_REQUESTED** [review]

**Comment by ionutgoldan** [comment]
Also, please elaborate a test coverage plan (just leave a comment in this PR).
After that, you can start implementing the test coverage.

**Comment by esanuandra** [comment]
> ~We should disable the dropdown if the user isn't logged in.~
> Oh, I just noticed this is disabled. Nevermind....
> 
> dropdown-shouldnt-be-visible-if-logout

I believe it is better if we disable the toggle and make it have a similar behavior like the left checkbox
image

**Comment by esanuandra** [comment]
### Tests:

1. user is not logged in -> check alerts dropdown toggle is disabled
2. user is logged in -> check alerts dropdown toggle is enabled
3. user is logged in & clicks on check alerts dropdown toggle -> check alerts dropdown is shown with the 4 options ('All', 'None', 'Triaged', 'Untriaged')
4. user is logged in & checks 'All' option from check alerts dropdown -> all visible alerts are checked
5. user is logged in & checks 'None' option from check alerts dropdown -> no alert is being checked or alerts are unchecked if they were checked before
6. user is logged in & checks 'Triaged' option from check alerts dropdown -> all the alerts that are visible and not untriaged are checked
7. user is logged in & checks 'Untriaged' option from check alerts dropdown -> all the alerts that are visible and untriaged are checked
8. user is logged in & double clicks on 'All' option from check alerts dropdown -> all visible alerts are (remain) checked, the button does not act as a toggle, it does not unchecks alerts

File: ui/perfherder/alerts/SelectAlertsDropdown.jsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      alertSummary,
+    } = this.props;
+    const { untriaged, triaged, all, none } = status;
+
+    return (
+      <React.Fragment>
+        <Label check className="pl-1">
+          <Input
+            data-testid={`alert summary ${alertSummary.id.toString()} checkbox`}
+            aria-labelledby={`alert summary ${alertSummary.id.toString()} title`}
```
Comment (ionutgoldan) [inline ui/perfherder/alerts/SelectAlertsDropdown.jsx:27]: What's the purpose of this `aria-labelledby`? I feel like it's not used properly.

File: ui/perfherder/alerts/SelectAlertsDropdown.jsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      allSelected,
+      alertSummary,
+    } = this.props;
+    const { untriaged, triaged, all, none } = status;
+
+    return (
+      <React.Fragment>
+        <Label check className="pl-1">
+          <Input
+            data-testid={`alert summary ${alertSummary.id.toString()} checkbox`}
```
Comment (ionutgoldan) [inline ui/perfherder/alerts/SelectAlertsDropdown.jsx:26]: I prefer we don't use `data-testid` on the frontend just to enhance our testing abilities.
react testing library has lots of ways of identifying test components. For example, by selecting them based on ARIA attributes.

File: ui/perfherder/alerts/AlertTable.jsx [hunk 5]
```diff
@@ -217,32 +240,21 @@ export default class AlertTable extends React.Component {
...
-                          issueTrackers={issueTrackers}
-                          user={user}
-                          updateAssignee={this.updateAssignee}
-                        />
-                      </Label>
+                    <FormGroup check className="d-inline-flex">
+                      <SelectAlertsDropdown
+                        selectAlertsByStatus={this.selectAlertsByStatus}
+                        user={user}
+                        alertSummary={alertSummary}
```
Comment (ionutgoldan) [inline ui/perfherder/alerts/AlertTable.jsx:247]: I don't think this parameter is that useful.

File: ui/perfherder/alerts/AlertTable.jsx [hunk 6]
```diff
@@ -176,6 +169,36 @@ export default class AlertTable extends React.Component {
     return { failureStatus };
   };
 
+  selectAlertsByStatus = (selectedStatus) => {
```
Comment (ionutgoldan) [inline ui/perfherder/alerts/AlertTable.jsx:172]: This logic feels more like the responsibility of `SelectAlertsDropdown`. Wouldn't it be better to try to move it there?
All the parent needs to know is what alerts have been selected by the dropdown & whether all have been selected or not.

**Review by ionutgoldan — CHANGES_REQUESTED** [review]

**Comment by ionutgoldan** [comment]
> 
> ### Tests:
> 
>     1. user is not logged in -> check alerts dropdown toggle is disabled
> 
>     2. user is logged in -> check alerts dropdown toggle is enabled
> 
>     3. user is logged in & clicks on check alerts dropdown toggle -> check alerts dropdown is shown with the 4 options ('All', 'None', 'Triaged', 'Untriaged')

This is redundant, as it is included in the 4 tests bellow.

>     4. user is logged in & checks 'All' option from check alerts dropdown -> all visible alerts are checked
> 
>     5. user is logged in & checks 'None' option from check alerts dropdown -> no alert is being checked or alerts are unchecked if they were checked before
> 
>     6. user is logged in & checks 'Triaged' option from check alerts dropdown -> all the alerts that are visible and not untriaged are checked
> 
>     7. user is logged in & checks 'Untriaged' option from check alerts dropdown -> all the alerts that are visible and untriaged are checked
> 
>     8. user is logged in & double clicks on 'All' option from check alerts dropdown -> all visible alerts are (remain) checked, the button does not act as a toggle, it does not unchecks alerts

I don't think this is needed.

File: ui/perfherder/alerts/SelectAlertsDropdown.jsx (hunk 4 above)
Comment (ionutgoldan) [reply ui/perfherder/alerts/SelectAlertsDropdown.jsx]: Seems like without this, we'd break some of our existing tests. I'm cancelling this request.

File: ui/perfherder/alerts/SelectAlertsDropdown.jsx [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+            onChange={() => {
+              return allSelected
+                ? this.selectAlertsByStatus(none)
+                : this.selectAlertsByStatus(all);
+            }}
+          />
+        </Label>
+        <UncontrolledDropdown size="sm" className="mr-2">
+          <DropdownToggle
+            aria-label="select-alerts-dropdown-toggle-label"
```
Comment (ionutgoldan) [inline ui/perfherder/alerts/SelectAlertsDropdown.jsx:72]: The string here should look like natural language. Also, we should make it shorter. Maybe "Alert selection options"?

File: tests/ui/perfherder/alerts-view/select_alerts_dropdown_test.jsx [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+test('Select alerts dropdown toggle is disabled when user is not staff', async () => {
+  const { getByLabelText } = selectAlertsDropdown(
+    testUser({ isLoggedIn: true, isStaff: false }),
+  );
+
+  const dropdownToggle = await waitFor(() =>
+    getByLabelText('select-alerts-dropdown-toggle-label'),
+  );
+
+  expect(dropdownToggle).toHaveClass('disabled');
```
Comment (ionutgoldan) [inline tests/ui/perfherder/alerts-view/select_alerts_dropdown_test.jsx:42]: There's a better suited test API for this: `toBeDisabled()`.

File: tests/ui/perfherder/alerts-view/select_alerts_dropdown_test.jsx [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+test('Select alerts dropdown toggle is enabled when user is staff', async () => {
+  const { getByLabelText } = selectAlertsDropdown(
+    testUser({ isLoggedIn: true, isStaff: true }),
+  );
+
+  const dropdownToggle = await waitFor(() =>
+    getByLabelText('select-alerts-dropdown-toggle-label'),
+  );
+
+  expect(dropdownToggle.classList.contains('disabled')).toBe(false);
```
Comment (ionutgoldan) [inline tests/ui/perfherder/alerts-view/select_alerts_dropdown_test.jsx:54]: Again, there's a better API for this: `toBeEnabled()`.

File: tests/ui/perfherder/alerts-view/select_alerts_dropdown_test.jsx [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+});
+
+test("Selecting 'All' option from dropdown checks all the visibile alerts", async () => {
```
Comment (ionutgoldan) [inline tests/ui/perfherder/alerts-view/select_alerts_dropdown_test.jsx:57]: Typo on "visibile".

File: tests/ui/perfherder/alerts-view/select_alerts_dropdown_test.jsx [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+import { alertStatusMap } from '../../../../ui/perfherder/constants';
+import { getStatus } from '../../../../ui/perfherder/helpers';
+
+const testUser = ({ isLoggedIn, isStaff }) => ({
+  username: 'mozilla-ldap/test_user@mozilla.com',
+  isLoggedIn,
+  isStaff,
+  email: 'test_user@mozilla.com',
+});
+const testAlertSummary = testAlertSummaries[0];
```
Comment (ionutgoldan) [inline tests/ui/perfherder/alerts-view/select_alerts_dropdown_test.jsx:15]: This summary only has untriaged alerts. We should make sure we have multiple alerts with multiple statuses. I think `untriaged`, `invalid` & `acknowledged` would suffice.

**Review by ionutgoldan — CHANGES_REQUESTED** [review]

**Review by ionutgoldan — APPROVED** [review]
Looks good!

**Pull request merged**